### PR TITLE
add whistle fake in simulator for opponent kickoff

### PR DESCRIPTION
**Pull request by MaikRe**
## Why? What?

We have no current method of simulating whistles in the behavior simulator.
The only transition into playing is through the `game_controller_state` `Playing`, which immediately sets `ball_is_free` to `true`. Testing midfielder or other positioning in the `ball_is_free = false` phase was impossible until this PR.

Fixes #

## How to Test

Run  behavior simulator with `golden_goal_opponent_kickoff` and observe that between frames 1700 and 1713 the ball is not free and thus the correct behavior from the midfielders can be observed.
If you then revert #1307 you can see the midfielders plan paths into the opponent half.

### CODE REVIEW

File: crates/code_generation/src/cyclers.rs [hunk 1]
```diff
@@ -124,7 +124,7 @@ fn generate_struct(cycler: &Cycler, cyclers: &Cyclers, mode: CyclerMode) -> Toke
     };
 
     quote! {
-        pub(crate) struct Cycler<HardwareInterface>  {
+        pub struct Cycler<HardwareInterface>  {
```
Comment (schluis) [inline crates/code_generation/src/cyclers.rs:127]: why this change?

File: crates/hulk_behavior_simulator/src/robot.rs [hunk 2]
```diff
@@ -83,6 +83,7 @@ impl Robot {
             .borrow_mut()
             .insert("additional_outputs".to_string());
 
+        database.main_outputs.buttons.is_chest_button_pressed_once = true;
```
Comment (schluis) [inline crates/hulk_behavior_simulator/src/robot.rs:86]: Please move this up to the other statements in this style, i.e. e.g. line 82

File: crates/hulk_behavior_simulator/src/simulator.rs [hunk 3]
```diff
@@ -174,6 +179,81 @@ impl Simulator {
...
+                        .robots
+                        .get_mut(&player_number)
+                        .unwrap()
+                        .database
+                        .main_outputs
+                        .ground_to_field
+                        .unwrap()
+                        .inner
+                        .translation
+                        .x;
```
Comment (schluis) [inline crates/hulk_behavior_simulator/src/simulator.rs:223]:
```suggestion
                        .y;
```

File: crates/hulk_behavior_simulator/src/simulator.rs [hunk 4]
```diff
@@ -174,6 +179,81 @@ impl Simulator {
...
+                    let time = self.state.lock().time_elapsed;
+                    self.state
+                        .lock()
+                        .robots
+                        .get_mut(&player_number)
+                        .unwrap()
+                        .database
+                        .main_outputs
+                        .filtered_whistle
+                        .last_detection = Some(UNIX_EPOCH + time);
```
Comment (schluis) [inline crates/hulk_behavior_simulator/src/simulator.rs:252]: Why + UNIX_EPOCH?

File: crates/hulk_behavior_simulator/src/state.rs [hunk 5]
```diff
@@ -205,17 +204,8 @@ impl State {
...
-            robot.database.main_outputs.primary_state =
-                match (robot.is_penalized, self.game_controller_state.game_state) {
-                    (true, _) => PrimaryState::Penalized,
-                    (false, GameState::Initial) => PrimaryState::Initial,
-                    (false, GameState::Standby) => PrimaryState::Standby,
-                    (false, GameState::Ready { .. }) => PrimaryState::Ready,
-                    (false, GameState::Set) => PrimaryState::Set,
-                    (false, GameState::Playing { .. }) => PrimaryState::Playing,
-                    (false, GameState::Finished) => PrimaryState::Finished,
-                };
```
Comment (schluis) [inline crates/hulk_behavior_simulator/src/state.rs:217]: Why removed?

**Review by schluis — CHANGES_REQUESTED** [review]

File: crates/code_generation/src/cyclers.rs (hunk 1 above)
Comment (MaikRe) [reply crates/code_generation/src/cyclers.rs:127]:
```sh
warning: type `cyclers::control::Cycler<Interfake>` is more private than the item `robot::Robot::cycler`
  --> crates/hulk_behavior_simulator/src/robot.rs:32:5
   |
32 |     pub cycler: Cycler<Interfake>,
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ field `robot::Robot::cycler` is reachable at visibility `pub`
   |
note: but type `cyclers::control::Cycler<Interfake>` is only usable at visibility `pub(crate)`
  --> /home/maik/TUHH/hulk/target/incremental/build/hulk_behavior_simulator-f49526b9b6800b96/out/generated_code.rs:41:9
   |
41 |         pub(crate) struct Cycler<HardwareInterface> {
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   = note: `#[warn(private_interfaces)]` on by default

warning: `hulk_behavior_simulator` (lib) generated 1 warning
```

File: crates/hulk_behavior_simulator/src/simulator.rs (hunk 4 above)
Comment (MaikRe) [reply crates/hulk_behavior_simulator/src/simulator.rs:252]: Because this is the way the robots store time in the simulator. Time is just a duration of the simulator since start, UNIX_EPOCH makes it a "proper" system time.

File: tests/behavior/golden_goal_opponent_kickoff.lua [hunk 6]
```diff
@@ -43,10 +43,30 @@ function on_cycle()
 
     if state.cycle_count == 1700 then
         -- TODO: simulate whistle
-        state.ball.velocity = { -3.0, 1.5 }
+        whistle(1)
```
[the remark this answers is not in the document]
Comment (MaikRe) [reply tests/behavior/golden_goal_opponent_kickoff.lua:46]: Yes

File: crates/hulk_behavior_simulator/src/state.rs (hunk 5 above)
Comment (MaikRe) [reply crates/hulk_behavior_simulator/src/state.rs:217]: This was a simulator internal fake for the `game_controller_state_filter'. We now properly run this filter as a full node so this fake is no longer needed.

File: crates/hulk_behavior_simulator/src/simulator.rs (hunk 3 above)
Comment (MaikRe) [reply crates/hulk_behavior_simulator/src/simulator.rs:223]: Woops

File: crates/hulk_behavior_simulator/src/robot.rs (hunk 2 above)
Comment (MaikRe) [reply crates/hulk_behavior_simulator/src/robot.rs:86]: Done

**Review by schluis — APPROVED** [review]
❤️
